examples/neutra.py

```diff
@@ -0,0 +1,111 @@
+import argparse
+
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from jax import jit, random
+from jax.config import config as jax_config
+from jax.experimental import optimizers
+import jax.numpy as np
+from jax.scipy.special import logsumexp
+from jax.tree_util import tree_map
+
+from numpyro.contrib.autoguide import AutoIAFNormal
+import numpyro.distributions as dist
+from numpyro.handlers import sample
+from numpyro.hmc_util import initialize_model
+from numpyro.mcmc import mcmc
+from numpyro.svi import elbo, svi
+from numpyro.util import fori_collect
+
+"""
+This example illustrates how to use a trained AutoIAFNormal autoguide to transform a posterior to a
+Gaussian-like one. The transform will be used to get better mixing rate for NUTS sampler.
+
+[1] Hoffman, M. et al. (2019), ["NeuTra-lizing Bad Geometry in Hamiltonian Monte Carlo Using Neural Transport"]
+    (https://arxiv.org/abs/1903.03704).
+"""
+
+
+def dual_moon_pe(x):
+    term1 = 0.5 * ((np.linalg.norm(x, axis=-1) - 2) / 0.4) ** 2
+    term2 = -0.5 * ((x[..., :1] + np.array([-2., 2.])) / 0.6) ** 2
+    return term1 - logsumexp(term2, axis=-1)
+
+
+def dual_moon_model():
+    x = sample('x', dist.Uniform(-10 * np.ones(2), 10 * np.ones(2)))
+    pe = dual_moon_pe(x)
+    sample('log_density', dist.Delta(log_density=-pe), obs=0.)
+
+
+def make_transformed_pe(potential_fn, transform, unpack_fn):
+    def transformed_potential_fn(z):
+        # NB: currently, intermediates for ComposeTransform is None, so this has no effect
+        # see https://github.com/pyro-ppl/numpyro/issues/242
+        u, intermediates = transform.call_with_intermediates(z)
+        logdet = transform.log_abs_det_jacobian(z, u, intermediates=intermediates)
+        return potential_fn(unpack_fn(u)) + logdet
+
+    return transformed_potential_fn
+
+
+def main(args):
+    jax_config.update('jax_platform_name', args.device)
+
+    # FIXME: I cann't run mcmc for this potential function
+    # print("Start vanilla HMC...")
+    # vanilla_samples = mcmc(100, 100, init_params=np.array([2., 0.]), potential_fn=dual_moon_pe)
```

Comment:
I added a FIXME here for further investigate. If this is a hard-to-sample potential_fn, then the current approach is doing well. Otherwise, there should be a problem in our implementation.

Reply:
I think this is more likely due to some assumption we are making during adaptation. I would expect HMC to at least explore one of the modes depending on initialization.

Reply:
Yup, that is likely the case. I’ll make a separate issue for this.